src/coprocessor/dag/expr/arithmetic.rs

```diff
@@ -197,11 +197,94 @@ impl FnCall {
+        if rhs == 0 {
+            return Ok(None);
+        }
+
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+
+        let res = match (lus, rus) {
+            (true, true) => Ok(((lhs as u64) / (rhs as u64)) as i64),
+            (false, false) => div_i64(lhs, rhs),
+            (false, true) => div_i64_with_u64(lhs, rhs as u64).map(|r| r as i64),
+            (true, false) => div_u64_with_i64(lhs as u64, rhs).map(|r| r as i64),
+        };
+        res.map(Some)
+    }
+
+    pub fn intdivide_decimal(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_decimal(ctx, row));
+        let overflow = Error::overflow("BIGINT", &format!("({} DIV {})", lhs, rhs));
+        match lhs.into_owned() / rhs.into_owned() {
+            Some(v) => match v {
+                Res::Ok(v) => match v.as_i64() {
+                    Res::Ok(v_i64) => Ok(Some(v_i64)),
+                    Res::Truncated(v_i64) => Ok(Some(v_i64)),
+                    Res::Overflow(_) => Err(overflow),
+                },
+                Res::Truncated(v) => Err(Error::Truncated(format!("{} truncated", v))),
+                Res::Overflow(_) => Err(overflow),
+            },
+            None => Ok(None),
+        }
+    }
+
+    pub fn mod_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try_opt!(self.children[0].eval_real(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_real(ctx, row));
+        if rhs == 0f64 {
+            return Ok(None);
+        }
+        let res = lhs % rhs;
+        Ok(Some(res))
+    }
+
+    pub fn mod_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_decimal(ctx, row));
+        let overflow = Error::overflow("DECIMAL", &format!("({} % {})", lhs, rhs));
+        match lhs.into_owned() % rhs.into_owned() {
+            Some(v) => match v {
+                Res::Ok(v) => Ok(Some(Cow::Owned(v))),
+                Res::Truncated(v) => Err(Error::Truncated(format!("{} truncated", v))),
+                Res::Overflow(_) => Err(overflow),
+            },
+            None => Ok(None),
+        }
+    }
+
+    pub fn mod_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_int(ctx, row));
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+        if rhs == 0 {
+            return Ok(None);
```

Comment:
ditto, eval and check rhs first like `tidb` did.

Reply:
okay, thanks